runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -147,15 +154,16 @@ public Token appendCheckpoint() {
      *  @param snapshotTimestamp snapshot at which the checkpoint is taken.
      */
     @VisibleForTesting
-    public Token appendCheckpoint(Token snapshotTimestamp) {
+    public Token appendCheckpoint(Token snapshotTimestamp, Long streamTail) {
         long start = System.currentTimeMillis();
 
-        // If snapshotTimestamp (stream tail) is a negative value
+        // If stream tail is a negative value
         // we need to enforce a hole to ensure the start log address
         // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
         // Having a negative value will prevent trimming from happening.
-        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+        if (!Address.isAddress(streamTail)) {
```

Comment:
The stream tail query and the read have to happen on the same epoch, otherwise the sequencer can regress in between and the purpose of the read is defeated. Imagine this scenario:
1. Tail query
2. read token from step 1
3. write checkpoint

The point of step #2 is to prevent the sequencer from regressing beyond the value returned in step 1, but if the sequencer actually regresses between step 1 and step 2, then step 2 becoems a point-read and can result writing a checkpoint before that address. As a consequence, data loss can happen.